services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,267 @@
+            pendingChunkedRequests.remove(message.getCorrelationId());
+            chunkedMessageIdsPerRequest.remove(message.getCorrelationId());
+        });
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void send(String topic, Message message) {
+        messageProducer.send(topic, message);
+    }
+
+    /**
+     * Processes messages that come back to NB topics, usually messages come as a response to some request.
+     * If this response is for pended request then such request will be marked as completed.
+     * Messages might be chunked and not chunked. If chunked we need to wait until we receive the last one
+     * and only then collect all responses and complete the request.
+     *
+     * @param message received message.
+     */
+    void onResponse(Message message) {
+        if (!isValid(message)) {
+            logger.warn("Skipping invalid message: {}", message);
+            return;
+        }
+
+        if (message instanceof ErrorMessage) {
+            ErrorMessage error = (ErrorMessage) message;
+            logger.error("Response message is error: {}", error);
+
+            completeWithError(error);
+        } else if (message instanceof InfoMessage) {
+            if (isChunkedPendingResponse(message)) {
+                processChunkedMessage((ChunkedInfoMessage) message);
+            } else if (pendingRequests.containsKey(message.getCorrelationId())) {
+                InfoMessage infoMessage = (InfoMessage) message;
+                CompletableFuture<InfoData> request = pendingRequests.remove(message.getCorrelationId());
+                if (request != null) {
+                    request.complete(infoMessage.getData());
+                }
+            } else {
+                logger.trace("Received non-pending message");
+            }
+        }
+    }
+
+    /**
+     * Performs searching and collecting all chunked messages into one chain if possible.
+     */
+    private synchronized void processChunkedMessage(ChunkedInfoMessage received) {
+        Set<String> associatedMessages = chunkedMessageIdsPerRequest.get(received.getCorrelationId());
+        if (!associatedMessages.add(received.getMessageId())) {
+            logger.debug("Skipping chunked message, it is already received: {}", received);
+            return;
+        }
+
+        List<ChunkedInfoMessage> chain = messagesChains.get(received.getCorrelationId());
+        chain.add(received);
+
+        if (chain.size() == received.getTotalMessages() || received.getTotalMessages() == 0) {
+            completeRequest(chain);
+            messagesChains.remove(received.getCorrelationId());
+        }
+    }
+
+    /**
+     * Completes pending request with received responses.
+     */
+    private void completeRequest(List<ChunkedInfoMessage> chain) {
+        if (CollectionUtils.isEmpty(chain)) {
+            throw new IllegalStateException("Chain of messages should not be empty");
```

Comment:
one of the previous checks is `chain.size() == received.getTotalMessages() || received.getTotalMessages() == 0)`... it allow to think that 0 is valid size for chain.

Even if 0 is not valid size - we must notify caller(complete future) with error, not just drop this exception in code that know nothing what to do with it. We can't extract data from chunks, because there is 0 of them... but we have a `received` message in method `processChunkedMessage`, so we can use it's correlationId to lookup future object.